File: programming 1/lab 6/src/main/java/server/commands/ExecuteScriptCommand.java
package server.commands;


import common.communication.Response;
import common.communication.ResponseStatus;
import server.managers.CommandManager;
import server.exceptions.*;

import java.io.*;
import java.util.EmptyStackException;
import java.util.Stack;

/**
 * Команда для исполнения скрипта из файла.
 */
public class ExecuteScriptCommand extends Command {
    private final CommandManager commandManager;
    private static Stack<String> scriptStack;
    public ExecuteScriptCommand(CommandManager commandManager) {
        super("execute_script file_name", "исполнить скрипт из файла");
        scriptStack = new Stack<>();
        this.commandManager = commandManager;
    }

    /**
     * Исполняет команды из указанного файла.
     * @param file файл скрипта.
     */
    @Override
    public Response execute(String file) {
        if (file == null) throw new IllegalArgumentException("Команда execute_script требует аргумент file_name\n");
        else {
            if (file.equals("history1") || file.equals("history2") || file.equals("history3")) {
                try (BufferedReader reader = new BufferedReader(new FileReader(file));
                     BufferedWriter writer = new BufferedWriter(new FileWriter(file + "_copy"))) {
                    String line;
                    String previousLine = null;
                    while ((line = reader.readLine()) != null) {
                        if (previousLine != null) {
                            writer.write(previousLine);
                            writer.newLine();
                        }
                        previousLine = line;
                    }
                    file += "_copy";
                } catch (IOException e) {
                    return new Response(ResponseStatus.ERROR,"Ошибка ввода-вывода: " + e.getMessage());
                }
            }

            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
                String line;
                if ((!(scriptStack.isEmpty()) && (scriptStack.peek().equals(file))) || scriptStack.contains(file))
                    throw new RecursionScriptException();
                scriptStack.push(file);
                while ((line = reader.readLine()) != null) {
                    commandManager.execute(line, reader);
                }
                if (scriptStack.peek().equals(file)) scriptStack.pop();
                else throw new RecursionScriptException();
                return new Response(ResponseStatus.SUCCESS, "Скрипт успешно выполнен!");
            } catch (FileNotFoundException e) {
                return new Response(ResponseStatus.ERROR, String.format("Файл %s не найден =(\n", file));
            } catch (SecurityException e) {
                return new Response(ResponseStatus.ERROR,"Нет прав доступа к файлу " + file + ": " + e.getMessage());
            } catch (IOException e) {
                return new Response(ResponseStatus.ERROR,"Ошибка ввода-вывода: " + e.getMessage());
            } catch (EmptyStackException | RecursionScriptException e) {
                return new Response(ResponseStatus.ERROR, "Рекурсивный скрипт!");
            } finally {
                if (file.equals("history1_copy") || file.equals("history2_copy") || file.equals("history3_copy"))
                    new File(file).delete();
            }
        }
    }
}

File: programming 1/lab 7/src/main/java/server/commands/ExecuteScriptCommand.kt
package server.commands

import common.communication.Response
import common.communication.ResponseStatus
import server.exceptions.RecursionScriptException
import server.managers.CommandManager
import java.io.*
import java.util.*


/**
 * Команда для исполнения скрипта из файла.
 */
class ExecuteScriptCommand(private val commandManager: CommandManager) :
    Command("execute_script file_name", "исполнить скрипт из файла") {
    private val scriptStack = Stack<String>()

    /**
     * Исполняет команды из указанного файла.
     * @param arg файл скрипта.
     */
    override fun execute(arg: String?): Response {
        var file = arg
        requireNotNull(file) { "Команда ${name.split(' ')[0]} требует аргумент ${name.split(' ')[1]}" }
        if (file == "history1" || file == "history2" || file == "history3") {
            try {
                BufferedReader(FileReader(file)).use { reader ->
                    BufferedWriter(FileWriter(file + "_copy")).use { writer ->
                        var line: String?
                        var previousLine: String? = null
                        while ((reader.readLine().also { line = it }) != null) {
                            if (previousLine != null) {
                                writer.write(previousLine)
                                writer.newLine()
                            }
                            previousLine = line
                        }
                        file += "_copy"
                    }
                }
            } catch (e: IOException) {
                return Response(ResponseStatus.ERROR, "Ошибка ввода-вывода: " + e.message)
            }
        }

        try {
            BufferedReader(FileReader(file)).use { reader ->
                var line: String?
                if ((!(scriptStack.isEmpty()) && (scriptStack.peek() == file)) || scriptStack.contains(file)) throw RecursionScriptException()
                scriptStack.push(file)
                while ((reader.readLine().also { line = it }) != null) {
                    commandManager.execute(line!!, reader)
                }
                if (scriptStack.peek() == file) scriptStack.pop()
                else throw RecursionScriptException()
                return Response(ResponseStatus.SUCCESS, "Скрипт успешно выполнен!")
            }
        } catch (e: FileNotFoundException) {
            return Response(ResponseStatus.ERROR, "Файл $file не найден =(\n")
        } catch (e: SecurityException) {
            return Response(ResponseStatus.ERROR, "Нет прав доступа к файлу $file: ${e.message}")
        } catch (e: IOException) {
            return Response(ResponseStatus.ERROR, "Ошибка ввода-вывода: ${e.message}")
        } catch (e: EmptyStackException) {
            return Response(ResponseStatus.ERROR, "Рекурсивный скрипт!")
        } catch (e: RecursionScriptException) {
            return Response(ResponseStatus.ERROR, "Рекурсивный скрипт!")
        } finally {
            if (file == "history1_copy" || file == "history2_copy" || file == "history3_copy") File(file).delete()
        }
    }
}

